### PR TITLE
LSTM, GRU and RNN implementation for ChainerX

### PR DESCRIPTION
**Pull request by dido1998**
This depends on #7807

### CODE REVIEW

**Comment by dido1998** [comment]
@asi1024, did you check why the native backward pass may be failing? On further tests locally, I found that the native backward pass was mostly only failing for the Relu version of `n_step_rnn` and `n_step_birnn`.

**Comment by asi1024** [comment]
@dido1998 There are some same commits in PR #7282. Which PR should be merged at first?

**Comment by dido1998** [comment]
On discussion with @niboshi , we had decided that #7282 and #7678 won't be merged and only this one will be merged instead of them. I will close those PRs.

File: chainerx_cc/chainerx/routines/connection.cc [hunk 1]
```diff
@@ -372,4 +375,73 @@ Array Linear(const Array& x, const Array& w, const absl::optional<Array>& b, uin
     return out_matrix.Reshape(out_shape);
 }
 
+std::vector<Array> _extract_gates(Array x) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/connection.cc]: Could you make fixes to C++ function names?
In C++ we use `CamelCase` without a prefix of `_`.

File: chainerx_cc/chainerx/routines/connection.cc [hunk 2]
```diff
@@ -372,4 +375,73 @@ Array Linear(const Array& x, const Array& w, const absl::optional<Array>& b, uin
...
+    StackVector<int64_t, kMaxNdim> shape_vec;
+    shape_vec.push_back(x.shape()[0]);
+    shape_vec.push_back(static_cast<int>(x.shape()[1] / 4));
+    shape_vec.push_back(4);
+    for (int i = 2; i < x.ndim(); i++) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/connection.cc]: Let's avoid `int64_t` or `size_t` in place of `int` or `uint`.

File: chainerx_cc/chainerx/backward_builder.cc [hunk 3]
```diff
@@ -154,12 +154,32 @@ RetainedInputToken BackwardBuilder::RetainInput(size_t input_index) {
     return {internal::GetArrayBody(gsl::at(inputs_, input_index))->GetParams(), input_index};
 }
 
+std::vector<RetainedInputToken> BackwardBuilder::RetainInput(std::vector<size_t> indices) {
+    std::vector<RetainedInputToken> token;
+    for (size_t i : indices) {
+        CHAINERX_ASSERT(i < inputs_.size());
+        input_retention_record_.Record(i);
+        token.push_back({internal::GetArrayBody(gsl::at(inputs_, i))->GetParams(), i});
```
Comment (niboshi) [inline chainerx_cc/chainerx/backward_builder.cc]: `push_back` might be fine but we use `emplace_back` consistently in other places.
Let's use `emplace_back`.

File: chainerx_cc/chainerx/cuda/cuda_device/rnn.cu [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+namespace {
+
+__global__ void InitGPUDataKer(float* data, int64_t numElements, float* value) {
+    int64_t tid = blockIdx.x * blockDim.x + threadIdx.x;
+    if (tid < numElements) {
+        data[tid] = value[tid];
+    }
+}
+
+void InitGPUData(float* data, int64_t numElements, float* value) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/cuda/cuda_device/rnn.cu]: Please treat acronyms like "GPU" as words: spell as `Gpu`.

File: chainerx_cc/chainerx/cuda/cuda_device/rnn.cu [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+#include "chainerx/routines/creation.h"
+#include "chainerx/routines/manipulation.h"
+#include "chainerx/shape.h"
+#include "chainerx/stack_vector.h"
+
+namespace chainerx {
+namespace cuda {
+namespace {
+
+__global__ void InitGPUDataKer(float* data, int64_t numElements, float* value) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/cuda/cuda_device/rnn.cu]: Use `snake_case` for variable names: `numElements` -> `num_elements`, `gridDim` -> `grid_dim`, `wDesc_` -> `w_desc_` etc.

File: chainerx_cc/chainerx/routines/n_step_rnn.cc [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+    Array h_bar_prev = split_w[2] + r * split_h[2];
+    Array h_bar = Tanh(h_bar_prev);
+    std::vector<Array> out{};
+    out.reserve(1);
+    Array f = (1 - z) * h_bar + z * h;
+    out.emplace_back(f);
+    return out;
+}
+
+std::vector<Array> _lstm(
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/n_step_rnn.cc]: Please fix the function name. Perhaps `LstmImpl`?

File: chainerx_cc/chainerx/routines/n_step_rnn.cc [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+#include "chainerx/routines/creation.h"
+#include "chainerx/routines/hyperbolic.h"
+#include "chainerx/routines/linalg.h"
+#include "chainerx/routines/manipulation.h"
+#include "chainerx/routines/misc.h"
+#include "chainerx/routines/reduction.h"
+#include "chainerx/routines/type_util.h"
+namespace chainerx {
+
+Array StackWeight(const std::vector<Array>& ws) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/n_step_rnn.cc]: Please put private functions into unnamed namespace. (`namespace {  }`)

**Comment by dido1998** [comment]
@niboshi could you review this, I have made the changes

**Comment by niboshi** [comment]
Thanks!
Jenkins, test this please

File: chainerx_cc/chainerx/routines/connection.cc [hunk 8]
```diff
@@ -65,8 +66,9 @@ std::vector<Array> ExtractGates(Array x) {
     Shape shape{shape_vec};
     Array x_r = Reshape(x, shape);
     std::vector<Array> gates = Split(x_r, 4, 2);
-    for (size_t i = 0; i < gates.size(); i++) {
-        gates[i] = Squeeze(gates[i]);
+    int index = 0;
+    for (auto& gate : gates) {
+        gates[index++] = Squeeze(gate);
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/connection.cc]: `gate = Squeeze(gate)`?

File: chainerx_cc/chainerx/routines/n_step_rnn.cc [hunk 9]
```diff
@@ -153,8 +153,8 @@ std::vector<std::vector<Array>> OneDirectionalLoop(
         *c = Reshape(*c, h_shape);
     }
     std::vector<Array> h_list;
-    for (size_t i = 0; i < xs.size(); i++) {
-        Array x_t = xs[i];
+    for (auto& x : xs) {
+        Array x_t = x;
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/n_step_rnn.cc]: You can just remove `x_t` and use `x` instead?

File: chainerx_cc/chainerx/routines/connection.cc (hunk 8 above)
Comment (dido1998) [reply chainerx_cc/chainerx/routines/connection.cc]: done

File: chainerx_cc/chainerx/routines/n_step_rnn.cc (hunk 9 above)
Comment (dido1998) [reply chainerx_cc/chainerx/routines/n_step_rnn.cc]: done

**Comment by niboshi** [comment]
Thanks!
Jenkins, test this please

File: chainerx_cc/chainerx/routines/n_step_rnn.cc [hunk 10]
```diff
@@ -0,0 +1,500 @@
...
+    out.emplace_back(h_list);
+
+    return out;
+}
+
+template <typename Impl>
+std::vector<std::vector<Array>> NStepRnnImpl(
+        Impl&& impl,
+        int64_t n_layers,
+        Array hx,
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/n_step_rnn.cc]: There are clant-tidy errors.

I think this can be `const Array&` and `std::move` can be removed.

Comment (dido1998) [reply chainerx_cc/chainerx/routines/n_step_rnn.cc]: I have made the changes. Could you check?

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Test failure is unrelated.
Thank you so much!

**Pull request merged**